### PR TITLE
use the composite prober with the channel

### LINKED ISSUE #3188: Use the composite prober for the channel
**Problem**
Once https://github.com/knative-sandbox/eventing-kafka-broker/issues/3176 is complete, we should use the composite prober to probe the channel, instead of the async prober we are currently using. 

To do this, you will need to update the `Prober` type in the `Reconciler` struct from `Prober` to `NewProber` in https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/channel/channel.go. Then, you will be able to do something similar to: https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/broker/controller.go#L105 in https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/channel/controller.go

You will also need to add code to handle rotating the CA certificates, which can be done similarly to https://github.com/knative-sandbox/eventing-kafka-broker/blob/59371d0edf02a4f42886c1ab333b9647085ac850/control-plane/pkg/reconciler/broker/controller.go#L127-L135 

Then, you must add/modify the appropriate event handler: https://github.com/knative-sandbox/eventing-kafka-broker/blob/59371d0edf02a4f42886c1ab333b9647085ac850/control-plane/pkg/reconciler/broker/controller.go#L152-L155

You would then also need to do something similar for channelv2.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?

**Exit Criteria**
The channel (v1 and v2) probes with the composite prober instead of the async prober, and resolves properly for all transport encryption flags.

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve? 1

**Additional context (optional)**
Add any other context about the feature request here.
- https://github.com/knative-sandbox/eventing-kafka-broker/pull/3178
- https://github.com/knative-sandbox/eventing-kafka-broker/issues/3141

### PR DESCRIPTION
**Pull request by Rahul-Kumar-prog**
Fixes #3188 

## Proposed Changes

- Update the channel controller to use the prober with tls if the transport encryption flag is set to strict.

**Release Note**

```release-note
The Channel now uses a tls-enabled prober if the transport encryption flag is set to strict.
```

**Docs**

### CODE REVIEW

**Review by Cali0707 — CHANGES_REQUESTED** [review]
This is a good start @Rahul-Kumar-prog! Can you look into why your unit tests are failing on this PR:
```
# knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/channel_test [knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/channel.test]
Error: control-plane/pkg/reconciler/channel/channel_test.go:2134:23: cannot use proberMock (variable of type prober.Prober) as prober.NewProber value in struct literal: prober.Prober does not implement prober.NewProber (wrong type for method Probe)
		have Probe("context".Context, prober.Addressable, prober.Status) prober.Status
		want Probe("context".Context, prober.NewAddressable, prober.Status) prober.Status
		
```

It looks to me like we need to update the unit tests as well, similar to what I did in https://github.com/knative-sandbox/eventing-kafka-broker/pull/3175/files in the `broker_test.go` file.

**Comment by Rahul-Kumar-prog** [comment]
@Cali0707 How can I identify this failing cause ?

**Comment by Cali0707** [comment]
@Rahul-Kumar-prog look at the changes I made to the `broker_test.go` file in https://github.com/knative-sandbox/eventing-kafka-broker/pull/3175/files

**Comment by Rahul-Kumar-prog** [comment]
@Cali0707 Do I have to make changes in all the files that you have changed? Or similarly? Because you make changes in 15 file for the same kind of issue?

**Comment by Cali0707** [comment]
@Rahul-Kumar-prog you just have to follow the changes I made to `broker_test.go` in that PR in `channel_test.go` here.

**Comment by Rahul-Kumar-prog** [comment]
@Cali0707 have a look.

**Review by Cali0707 — COMMENTED** [review]
@Rahul-Kumar-prog thanks for your work so far! 

If you look at the issue, it also mentions implementing this for channel v2. Do you think you could make the same changes you have made here for channel v2?

File: control-plane/pkg/reconciler/channel/v2/controllerv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/controllerv2.go:126]:
```suggestion
		consumergroup.Enqueue("kafkachannel", impl.EnqueueKey)
```
If you look at the behaviour before, we were actually calling the `consumergroup.Enqueue` method, not a global resync method. I think that since all we are trying to do is also rotate the CA certs, we should call the original method here, rather than a new global resync method as otherwise we are changing the existing behaviour

File: control-plane/pkg/reconciler/channel/v2/controllerv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/controllerv2.go:117]:
```suggestion

```
See my other comment, I don't think we need this new method

**Review by Cali0707 — CHANGES_REQUESTED** [review]
Thanks for adding the channelv2 changes @Rahul-Kumar-prog !

File: control-plane/pkg/reconciler/channel/channel.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/channel.go:431]:
```suggestion
			Address:   &address,
			Addresses: []duckv1.Addressable{address},
```

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/channelv2.go:424]:
```suggestion
			Address:   &address,
			Addresses: []duckv1.Addressable{address},
```

**Review by Cali0707 — CHANGES_REQUESTED** [review]
@Rahul-Kumar-prog sorry I missed this earlier, but you need to add the address into the `Addresses` field for all of the `duckv1.AddressStatus` fields you are creating yourself.

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/channelv2.go:425]:
```suggestion
			Addresses: []duckv1.Addressable{address},
```

**Review by Cali0707 — CHANGES_REQUESTED** [review]

**Comment by Rahul-Kumar-prog** [comment]
Sorry I did very silly mistakes sorry.

**Comment by Cali0707** [comment]
@Rahul-Kumar-prog could you look into why the tests are failing?

**Comment by Leo6Leo** [comment]
@Rahul-Kumar-prog 

Sharing some tricks on how to debug rekt based reconciler test:
 ```
SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run <The test name that you would like to run>  -parallel=12 -timeout=30m ./test/rekt
```

Take one of your failed test as an example, 
```
 SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run TestEventTransformationForTrigger  -parallel=12 -timeout=30m ./test/e2e_new/
```

This command allows you to run **single** rekt test. You don't have to run all the rekt test and wait for hours for them to be finished.

**Comment by Rahul-Kumar-prog** [comment]
> @Rahul-Kumar-prog
> 
> Sharing some tricks on how to debug rekt based e2e test:
> 
> ```
> SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run <The test name that you would like to run>  -parallel=12 -timeout=30m ./test/rekt
> ```
> 
> Take one of your failed test as an example,
> 
> ```
>  SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run TestEventTransformationForTrigger  -parallel=12 -timeout=30m ./test/e2e_new/
> ```
> 
> This command allows you to run **single** rekt test. You don't have to run all the rekt test and wait for hours for them to be finished.

This command is not working.

**Comment by Leo6Leo** [comment]
> > @Rahul-Kumar-prog
> > Sharing some tricks on how to debug rekt based e2e test:
> > ```
> > SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run <The test name that you would like to run>  -parallel=12 -timeout=30m ./test/rekt
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Take one of your failed test as an example,
> > ```
> >  SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run TestEventTransformationForTrigger  -parallel=12 -timeout=30m ./test/e2e_new/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This command allows you to run **single** rekt test. You don't have to run all the rekt test and wait for hours for them to be finished.
> 
> This command is not working.

What is the error that saying it is not working? @Rahul-Kumar-prog

**Comment by Cali0707** [comment]
> > @Rahul-Kumar-prog
> > Sharing some tricks on how to debug rekt based e2e test:
> > ```
> > SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run <The test name that you would like to run>  -parallel=12 -timeout=30m ./test/rekt
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Take one of your failed test as an example,
> > ```
> >  SYSTEM_NAMESPACE=knative-eventing go test -v -tags=e2e -count=1 -run TestEventTransformationForTrigger  -parallel=12 -timeout=30m ./test/e2e_new/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This command allows you to run **single** rekt test. You don't have to run all the rekt test and wait for hours for them to be finished.
> 
> This command is not working.

@Rahul-Kumar-prog did you follow the steps in `DEVELOPMENT.md` to set up a cluster with eventing and eventing-kafka-broker installed and running?

**Comment by Rahul-Kumar-prog** [comment]
Screenshot 2023-08-17 at 10 03 46 PM

Getting Test ok and working fine @Cali0707 @Leo6Leo

**Comment by Leo6Leo** [comment]
> Screenshot 2023-08-17 at 10 03 46 PM
> 
> Getting Test ok and working fine @Cali0707 @Leo6Leo

@Rahul-Kumar-prog It said "no tests to run", which means it doesn't run anything. The name of the test you have in the picture is wrong.

**Comment by Cali0707** [comment]
Hey @Rahul-Kumar-prog any progress with debugging the test failures here? Is there anything I can help with?

**Comment by Cali0707** [comment]
@Rahul-Kumar-prog can you rebase this PR to have the newest version of the changes on `main`? I think that will likely fix your failing tests, as there have been a few bug fixes on the prober since you opened this PR

**Comment by Rahul-Kumar-prog** [comment]
can you please retest ?

File: control-plane/pkg/reconciler/channel/controller.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/controller.go:102]: @Rahul-Kumar-prog you need to keep the port string the same here:
```suggestion
	reconciler.Prober, err = prober.NewComposite(ctx, "", configs.IngressPodTlsPort, IPsLister, impl.EnqueueKey, &caCerts)
```
@pierDipi you changed the http port to `""` in https://github.com/knative-extensions/eventing-kafka-broker/pull/2164 - do you know if the TLS port should also be `""` on this PR?

File: control-plane/pkg/reconciler/channel/v2/controllerv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/controllerv2.go:110]:
```suggestion
	reconciler.Prober, err = prober.NewComposite(ctx, "", configs.IngressPodTlsPort, IPsLister, impl.EnqueueKey, &caCerts)
```
@pierDipi same question as above

**Review by Cali0707 — CHANGES_REQUESTED** [review]
@Rahul-Kumar-prog I found a few more places in your code which are causing some of the failing tests, if you make the changes I suggested more (hopefully all) of the tests will pass. The question I had for @pierDipi won't block any current tests from passing, but may affect whether this works when we add tls tests for the channel

File: control-plane/pkg/reconciler/channel/controller.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/channel/controller.go:102]: yes, the reason for this is that we probing the service of the Kafka channel which for http will use 80 and for https will use 443 which are both the defaults port

File: control-plane/pkg/reconciler/channel/controller.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/controller.go:102]:
```suggestion
	reconciler.Prober, err = prober.NewComposite(ctx, "", "", IPsLister, impl.EnqueueKey, &caCerts)
```

File: control-plane/pkg/reconciler/channel/v2/controllerv2.go
Comment (Cali0707) [inline control-plane/pkg/reconciler/channel/v2/controllerv2.go:110]:
```suggestion
	reconciler.Prober, err = prober.NewComposite(ctx, "", "", IPsLister, impl.EnqueueKey, &caCerts)
```

**Review by Cali0707 — CHANGES_REQUESTED** [review]
@Rahul-Kumar-prog @pierDipi per earlier conversation, these changes will set the port correctly for the case when we want to probe with TLS

**Comment by Cali0707** [comment]
Btw @Rahul-Kumar-prog, you normally don't need to merge the main branch into your PR unless you:
1. want to use new code from the main branch
2. Have a merge conflict
Otherwise, Prow will handle that for you once the PR is approved

**Comment by Rahul-Kumar-prog** [comment]
@Cali0707 thank you for telling I was not aware of that ?

**Review by Cali0707 — APPROVED** [review]
/lgtm
/approve
/hold in case @pierDipi wants any changes

**Pull request merged**
